Subgradients.thy

```diff
@@ -0,0 +1,371 @@
+        also have "abs (inner (u-v) x) \<le> norm (u-v) * norm x"
+          using Cauchy_Schwarz_ineq2 by blast
+        finally show ?thesis
+          by (simp add: mult.commute)
+      qed
+    next
+      case False
+      then show ?thesis
+      proof (cases "1 - y * inner v x \<le> 0")
+        case True
+        have "?f u \<ge> 0" by auto
+        have "abs(?f u - ?f v) = abs (?f u)" using  `1 - y * inner v x \<le> 0` by auto
+        also have "abs (?f u) \<le> abs (?f u - 1 + y * inner v x)" using False `1 - y * inner v x \<le> 0`
+          using `?f u \<ge> 0` by linarith
+        also have "abs (?f u - 1 + y * inner v x) \<le> abs (- y * inner v x + y * inner u x)" 
+          using False by auto
+        also have " abs (- y * inner v x + y * inner u x) = abs ( y * (inner u x - inner v x))"
+          by (metis add.commute minus_mult_commute ring_normalization_rules(2) vector_space_over_itself.scale_right_distrib)
+        also have " abs ( y * (inner u x - inner v x)) =  abs ( y * (inner (u-v) x))" 
+          by (simp add: inner_diff_left)
+        also have "abs ( y * (inner (u-v) x)) = abs y * abs (inner (u-v) x)" 
+          using abs_mult by blast
+        also have "abs y * abs (inner (u-v) x) = abs (inner (u-v) x)" using assms by auto
+        also have "abs (inner (u-v) x) \<le> norm (u-v) * norm x"
+          using Cauchy_Schwarz_ineq2 by blast
+        finally show ?thesis
+          by (simp add: mult.commute)
+      next
+        case False
+        have "abs(?f u - ?f v)  \<le> abs (?f u - 1 + y * inner v x)"
+          by simp
+        also have "abs (?f u - 1 + y * inner v x) \<le> abs (- y * inner v x + y * inner u x)" 
+          using False by auto
+        also have " abs (- y * inner v x + y * inner u x) = abs ( y * (inner u x - inner v x))"
+          by (metis add.commute minus_mult_commute ring_normalization_rules(2) vector_space_over_itself.scale_right_distrib)
+        also have " abs ( y * (inner u x - inner v x)) =  abs ( y * (inner (u-v) x))" 
+          by (simp add: inner_diff_left)
+        also have "abs ( y * (inner (u-v) x)) = abs y * abs (inner (u-v) x)" 
+          using abs_mult by blast
+        also have "abs y * abs (inner (u-v) x) = abs (inner (u-v) x)" using assms by auto
+        also have "abs (inner (u-v) x) \<le> norm (u-v) * norm x"
+          using Cauchy_Schwarz_ineq2 by blast
+        finally show ?thesis
+          by (simp add: mult.commute)
+      qed
+    qed
+  qed
+  then show ?thesis unfolding lipschitz_on_def
+    by (simp add: dist_norm)
+qed
+
+lemma sum_of_max:
+  fixes a :: real
+  fixes b :: real
+  shows "max 0 a + max 0 b \<ge>  max 0 (a+b)"
+  by simp
+
+
+lemma hinge_loss_convex:
+  fixes H :: "'a::euclidean_space set"
+    and D :: "('a \<times> real) pmf"
+  shows "\<forall>z \<in> D. convex_on H (\<lambda> h. hinge_loss h z)"
+proof
+  fix z
+  assume "z \<in> D"
+  show " convex_on H (\<lambda> h. hinge_loss h z)" unfolding hinge_loss.simps
+  proof(rule)+
```

Comment:
what about the following, and then using `a` and `b` instead of `fst z`and `snd z`
```suggestion
lemma hinge_loss_alt: "hinge_loss w (a,b) = (max 0 (1 - b * inner w a))"
  by simp

lemma hinge_loss_convex:
  fixes H :: "'a::euclidean_space set"
    and D :: "('a × real) pmf" 
  assumes "(a,b)∈D"
  shows "convex_on H (λ h. hinge_loss h (a,b))"
  unfolding hinge_loss_alt 
proof (rule)+
```